### PR TITLE
refactor: introduce new `Output` with `OutputMeta`

### LINKED ISSUE #3457: Refactor `Output` for carrying output metadata
### What type of enhancement is this?

Refactor

### What does the enhancement do?

`Output` struct now
```Rust
pub enum Output {
    AffectedRows(usize),
    RecordBatches(RecordBatches),
    Stream(SendableRecordBatchStream, Option<Arc<dyn PhysicalPlan>>),
}
```

Currently `Output` is merely a wrap-up for actual output data. It is difficult and inappropriate to add some meta data (for example metrics data produced during the insert or query) with this struct.

Here we propose a new struct in order to fix this situation.

```Rust
pub struct Output {
    data: OutputData,
    meta: OutputMeta,
}

pub enum OutputData {
    AffectedRows(usize),
    RecordBatches(RecordBatches),
    Stream(SendableRecordBatchStream),
}

pub enum OutputMeta {
    plan: Option<Arc<dyn PhysicalPlan>>,
    cost: usize,
}
```

### Implementation challenges

The original `Output` is widely used across the project. It might be a lot of work to finish this refactor.

### PR DESCRIPTION
**Pull request by shuiyisong**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
closes #3457

## What's changed and what's your intention?

This pr refactors `Output`, turning the old one into `OutputData` and adding new `OutputMeta`. Since it's widely used across the project, we try to minimize the change by focusing on only changing the struct. Related feature addition will be made in a separate pr.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/common/query/src/lib.rs [hunk 1]
```diff
@@ -30,38 +30,86 @@ pub mod prelude;
...
     AffectedRows(usize),
     RecordBatches(RecordBatches),
-    Stream(SendableRecordBatchStream, Option<Arc<dyn PhysicalPlan>>),
+    Stream(SendableRecordBatchStream),
+}
+
+/// OutputMeta stores meta information produced/generated during the execution
+#[derive(Debug, Default)]
+pub struct OutputMeta {
+    pub plan: Option<Arc<dyn PhysicalPlan>>,
```
Comment (waynexia) [inline src/common/query/src/lib.rs:51]:
```suggestion
    /// May exist for query output. One can retrieve execution metrics from this plan.
    pub plan: Option<Arc<dyn PhysicalPlan>>,
```

**Review by waynexia — APPROVED** [review]
I've skimmed this patch and it looks good to me. Thanks for undertaking this overwhelming challenge 💯

File: src/common/query/src/lib.rs [hunk 2]
```diff
@@ -30,38 +30,87 @@ pub mod prelude;
...
-    pub fn new_stream(stream: SendableRecordBatchStream) -> Self {
-        Output::Stream(stream, None)
+    pub fn new_with_affectedrows(affected_rows: usize) -> Self {
+        Self {
+            data: OutputData::AffectedRows(affected_rows),
+            meta: Default::default(),
+        }
+    }
+
+    pub fn new_with_recordbatches(recordbatches: RecordBatches) -> Self {
```
Comment (evenyag) [inline src/common/query/src/lib.rs:64]: nitpick
```suggestion
    pub fn new_with_affected_rows(affected_rows: usize) -> Self {
        Self {
            data: OutputData::AffectedRows(affected_rows),
            meta: Default::default(),
        }
    }

    pub fn new_with_record_batches(recordbatches: RecordBatches) -> Self {
```

**Review by evenyag — APPROVED** [review]
Mostly LGTM.

**Pull request merged**